### PR TITLE
docs: Document applications in any namespace

### LINKED ISSUE #2785: Locking "App Of Apps" to a single project and namespace
# Summary

In AppProject we can limit Applications' possibilities using destinations, clusterResourceWhitelist and namespaceResourceBlacklist. For example, an Application can create any resources, as long as they are created within a particular namespace. However this does not allow creation of new Applications (App of Apps) because only Applications from `argocd` namespace are being managed. 

If we allow creation of Applications in 'argocd`, then other resources can be created there too. Pods and Jobs get access to a very powerful service account of ArgoCD.

# Motivation

We'd like to give development teams freedom to manage Applications, however 
* they should belong to the same ArgoCD project and 
* all applications' resources should be created in the same kubernetes namespace.

# Proposal

I would consider 
* letting ArgoCD to manage Applications, which are created by other Applications, regardless of their namespaces and 
* (optionally) enforcing sub-apps to inherit project from the parent

### PR DESCRIPTION
**Pull request by jannfis**
Preview: https://argo-cd--10678.org.readthedocs.build/en/10678/operator-manual/app-any-namespace/

Closes https://github.com/argoproj/argo-cd/issues/2785

Signed-off-by: jannfis <jann@mistrust.net>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: docs/operator-manual/app-any-namespace.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
+# Applications in any namespace
+
+**Current feature state**: Beta
+
+!!! warning
+    Please read this documentation carefully before you enable this feature. Misconfiguration could lead to potential security issues.
+
+## Introduction
+
+As of version 2.5, Argo CD supports managing `Application` resources in namespaces other than the control plane's namespace (which is usally `argocd`), but this feature has to be explicitly enabled and configured appropriately.
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:10]: usually

File: docs/operator-manual/app-any-namespace.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+Argo CD administrators can define a certain set of namespaces where `Application` resources may be created, updated and reconciled in. However, applications in these additional namespaces will only be allowed to use certain `AppProjects`, as configured by the Argo CD administrators. This allows ordinary Argo CD users (e.g. application teams) to use patterns like declarative management of `Application` resources, implementing app-of-apps and others without the risk of a privilege escalation through usage of other `AppProjects` that would exceed the permissions granted to the application teams.
+
+Some manual steps will need to be performed by the Argo CD administrator in order to enable this feature. 
+
+!!! note
+    This feature is considered beta as of now. Some of the implementation details may change over the course of time until it is promoted to a stable status. We will be happy about early adopters and to receive bug reports and feedback.
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:17]: We will be happy if early adopters use this feature and provide us with bug reports and feedback?

File: docs/operator-manual/app-any-namespace.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+To enable annotation based resource tracking, refer to the documentation about [resource tracking methods](../../user-guide/resource_tracking/)
+
+## Implementation details
+
+### Overview
+
+In order for an application to be managed and reconciled outside the Argo CD's control plane namespace, two prerequisites must match:
+
+1. The `Application`'s namespace must be explicitly enabled using the `--application-namespaces` parameter for the `argocd-application-controller` and `argocd-server` workloads. This parameter controls the list of namespaces that Argo CD will be allowed to source `Application` resources from globally. Any namespace not configured here cannot be used from any `AppProject`.
+1. The `AppProject` referenced by the `.spec.project` field of the `Application` must have the namespace listed in its `.spec.sourceNamespaces` field. This setting will determine whether an `Application` may use a certain `AppProject`. If an `Application` specifies an `AppProject` that it is not allowed to, Argo CD refuses to process this `Application`. As stated above, any namespace configured in the `.spec.sourceNamespaces` field must also be enabled globally.
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:38]: How about just simply: "If an Application specifies an AppProject that is not allowed, then Argo CD refuses to...."

File: docs/operator-manual/app-any-namespace.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+`Applications` in different namespace can be created and managed just like any other `Application` in the `argocd` namespace previously, either declaratively or through the Argo CD API (e.g. using the CLI, the web UI, the REST API, etc).
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:40]: Applications in a different namespace can be created and managed....

File: docs/operator-manual/app-any-namespace.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+### Reconfigure Argo CD to allow certain namespaces
+
+#### Change workload startup parameters
+
+In order to enable this feature, the Argo CD administrator must reconfigure the `argocd-server` and `argocd-application-controller` workloads to add the `--application-namespaces` parameter to the container's startup command.
+
+The `--application-namespaces` parameter takes a comma-separated list of namespaces where `Applications` are to be allowed in. Each entry of the list supports shell-style wildcards such as `*`, so for example the entry `app-team-*` would match `app-team-one` and `app-team-two`. To enable all namespaces on the cluster where Argo CD is running on, you can just `*`, i.e. `--application-namespaces=*`.
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:48]: you can just use '*'......

File: docs/operator-manual/app-any-namespace.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+The startup parameters for both, the `argocd-server` and the `argocd-application-controller` can also be conviently set up and kept in sync by specifying the `application.namespaces` settings in the `argocd-cmd-params-cm` ConfigMap _instead_ of changing the manifests for the respective workloads. For example:
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:50]: conveniently

File: docs/operator-manual/app-any-namespace.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+```
+
+!!! note
+    At some later point in time, we may make this cluster role part of the default installation manifests.
+
+### Allowing additional namespaces in an AppProject
+
+Any user with Kubernetes access to the Argo CD control plane's namespace (`argocd`), especially those with permissions to create or update `Applications` in a declarative way, is to be considered an Argo CD admin.
+
+This prevents unprivileged Argo CD users from declaratively creating or managing `Applications` in the past, and they were constrained to using the API instead with Argo CD RBAC ensuring they could only create `Applications` in `AppProjects` they were allowed to.
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:81]: I think this long sentence could probably be split into two.  Maybe just break it at:
" .... in the past.  The users were constrained to....."

Also, what about instead?

...ensuring they could create Applications in only those AppProjects they had sufficient permissions?

or

...ensuring they could create Applications in only allowable AppProjects as defined by the RBAC rules?


Or simply

...ensuring they could create Applications in only allowable AppProjects.

File: docs/operator-manual/app-any-namespace.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+argocd app manifests foo/bar
+```
+
+As stated previously, for applications in the Argo CD's control plane namespace, you can omit the namespace from the application name.
+
+### Using the UI
+
+Similar to the CLI, you can refer to the application in the UI as `foo/bar`.
+
+For example, to create an application named `bar` in the namespace `foo` in the web UI, set the application name in the creation input mask to `foo/bar`. If the namespace is omitted, the control plane's namespace will be used.
```
Comment (keithchong) [inline docs/operator-manual/app-any-namespace.md:208]: "in the creation input mask" --> Are you referring to the Application Name field or something else?

Screen Shot 2022-12-15 at 10 33 40 AM

File: docs/operator-manual/app-any-namespace.md (hunk 1 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Fixed, thanks!

File: docs/operator-manual/app-any-namespace.md (hunk 2 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Good suggestion, applied. Thanks!

File: docs/operator-manual/app-any-namespace.md (hunk 3 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Reworded a little. Thanks.

File: docs/operator-manual/app-any-namespace.md (hunk 4 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Chose the latter. Thanks.

File: docs/operator-manual/app-any-namespace.md (hunk 5 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: :+1: , thanks.

File: docs/operator-manual/app-any-namespace.md (hunk 6 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Fixed, thanks.

File: docs/operator-manual/app-any-namespace.md (hunk 7 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Reworded, thanks!

File: docs/operator-manual/app-any-namespace.md (hunk 8 above)
Comment (jannfis) [reply docs/operator-manual/app-any-namespace.md]: Clarified the proper field, thanks.

**Review by keithchong — APPROVED** [review]
LGTM

**Comment by crenshaw-dev** [comment]
Cherry-picked onto release-2.5 for 2.5.6. Thanks, y'all!

**Comment by daanvinken** [comment]
@crenshaw-dev / @jannfis sorry for tagging you both directly. We are quite invested on using applications and appsets in different namespaces. And my question is related to argoproj/argo-cd/issues/#18693.

Regarding:
> At some later point in time, we may make this cluster role part of the default installation manifests. 

Do you perhaps know whether we want to include this or perhaps are you aware of a different place where this is being discussed?

In my understanding the application-controller already has access to secrets/configmap (basically any resource), so do we view the applicationset-controller as a different attack vector / risk?

I am happy to contribute

**Pull request merged**